relay-dynamic-config/src/metrics.rs

```diff
@@ -170,13 +170,112 @@ impl TransactionMetricsConfig {
     }
 }
 
+/// Combined view of global and project-specific metrics extraction configs.
+#[derive(Debug, Clone)]
+pub struct CombinedMetricExtractionConfig<'a> {
+    global: &'a MetricExtractionTemplates,
+    project: &'a MetricExtractionConfig,
+}
+
+impl<'a> CombinedMetricExtractionConfig<'a> {
+    /// Creates a new combined view from two references.
+    pub fn new(global: &'a MetricExtractionTemplates, project: &'a MetricExtractionConfig) -> Self {
+        Self { global, project }
+    }
+
+    /// Returns an iterator of metric specs.
+    pub fn metrics(&self) -> impl Iterator<Item = &MetricSpec> {
+        let project = self.project.metrics.iter();
+        let enabled_global = self
+            .enabled_templates()
+            .flat_map(|template| template.metrics.iter());
+
+        project.chain(enabled_global)
+    }
+
+    /// Returns an iterator of tag mappings.
+    pub fn tags(&self) -> impl Iterator<Item = &TagMapping> {
+        let project = self.project.tags.iter();
+        let enabled_global = self
+            .enabled_templates()
+            .flat_map(|template| template.tags.iter());
+
+        project.chain(enabled_global)
+    }
+
+    fn enabled_templates(&self) -> impl Iterator<Item = &MetricExtractionTemplate> {
+        self.global.templates.iter().filter_map(|(key, template)| {
+            let is_enabled_by_override =
+                self.project.global_templates.get(key).map(|c| c.is_enabled);
+            let is_enabled = is_enabled_by_override.unwrap_or(template.is_enabled);
+
+            is_enabled.then_some(template)
+        })
+    }
+}
+
+impl<'a> From<&'a MetricExtractionConfig> for CombinedMetricExtractionConfig<'a> {
+    fn from(value: &'a MetricExtractionConfig) -> Self {
+        Self::new(MetricExtractionTemplates::EMPTY, value)
+    }
+}
+
+/// Global templates for metric extraction.
+///
+/// Templates can be enabled or disabled by project configs.
+#[derive(Clone, Default, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MetricExtractionTemplates {
+    /// Mapping from template name to metrics specs & tags.
+    #[serde(skip_serializing_if = "BTreeMap::is_empty")]
+    pub templates: BTreeMap<String, MetricExtractionTemplate>,
+}
+
+impl MetricExtractionTemplates {
+    const EMPTY: &'static Self = &Self {
+        templates: BTreeMap::new(),
+    };
+
+    /// Returns `true` if the continaed templates are empty.
+    pub fn is_empty(&self) -> bool {
+        self.templates.is_empty()
+    }
+}
+
+/// Set of metrics & tags that can be enabled or disabled as a group.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct MetricExtractionTemplate {
```

Comment:
Not sure if we should call this "templates", "sets", or something else. Bikeshedding welcome!

Reply:
I don't like "templates" because it kind of implies it needs something else to be "complete" and it's not useful on its own, which is not the case. I can't think of a name I really like, but "sets", "specs" or "definitions" sound better (or "metric tagging config"?).

Reply:
I like `Template` but `group` could also work. You can enable/disable certain groups. I think we also use the groups naming already throughout Relay.

Reply:
`group` it is! Against `template` you could argue that it's not incomplete, against `set` you could argue that it's not a set in the mathematical sense. "group" is neutral (even though it's not a group in the mathematical sense either :D).